File: src/information_gain.py
#! /usr/bin/python
import numpy as np
"""
    n, p might represent left and right
    1, 2 might represent false and true
"""

def information_gain(n1, n2, p1, p2):
    n1p1 = float(n1) + p1
    n2p2 = float(n2) + p2
    n = float(n1) + n2
    p = float(p1) + p2
    pn = float(n) + p

    q1 = n/pn
    q2 = p/pn
    q3 = n1/n1p1
    q4 = p1/n1p1
    q5 = n2/n2p2
    q6 = p2/n2p2

    current = 0.
    rem1 = 0.
    rem2 = 0.

    eps = 0.

    if q1 > eps:
        current += q1*np.log2(q1)
    if q2 > eps:
        current += q2*np.log2(q2)
    current *= -1.

    if n1p1 > eps:
        if q3 > eps:
            rem1 += q3*np.log2(q3)
        if q4 > eps:
            rem1 += q4*np.log2(q4)
        rem1 *= -n1p1/pn

    if n2p2 > eps:
        if q5 > eps:
            rem2 += q5*np.log2(q5)
        if q6 > eps:
            rem2 += q6*np.log2(q6)
        rem2 *= -n2p2/pn

    remainder = rem1 + rem2

    return current - remainder

File: src/information_gain.cc
#include <ctgmath>

double information_gain (int n1, int n2, int p1, int p2)
{
    double n1p1 = n1 + p1;
    double n2p2 = n2 + p2;
    double n = n1 + n2;
    double p = p1 + p2;

    double pn = n + p;

    double current = 0.;
    double rem1 = 0.;
    double rem2 = 0.;

    double eps = 0.;

    if (pn > eps) {
        double q1 = n/pn;
        double q2 = p/pn;
        if (q1 > eps) {
            current += q1 * std::log2(q1);
        }
        if (q2 > eps) {
            current += q2 * std::log2(q2);
        }
        current *= -1.;
    } else {
        return 0.;
    }

    if (n1p1 > eps) {
        double q3 = n1/n1p1;
        double q4 = p1/n1p1;

        if (q3 > eps) {
            rem1 += q3 * std::log2(q3);
        }
        if (q4 > eps) {
            rem1 += q4 * std::log2(q4);
        }
        rem1 *= -n1p1/pn;
    }

    if (n2p2 > eps) {
        double q5 = n2/n2p2;
        double q6 = p2/n2p2;

        if (q5 > eps) {
            rem2 += q5 * std::log2(q5);
        }
        if (q6 > eps) {
            rem2 += q6 * std::log2(q6);
        }
        rem2 *= -n2p2/pn;
    }

    double remainder = rem1 + rem2;

    return current - remainder;
}
